4/m_recaptcha.cpp

```diff
@@ -0,0 +1,230 @@
+    {
+        if (!sha256)
+        {
+            throw ModuleException(this, "*** reCAPTCHA: Cannot initialize - SHA256 module is required but not loaded.");
+        }
+    }
+
+    void ReadConfig(ConfigStatus& status) override
+    {
+        auto& tag = ServerInstance->Config->ConfValue("captchaconfig");
+
+        // SQL query
+        captcha_url = tag->getString("url");
+        query = tag->getString("query", "SELECT COUNT(*) FROM recaptcha_app_verificationtoken WHERE token = $1 AND created_at + INTERVAL '30 minutes' > NOW()", 1);
+        
+        // HMAC key for hashing
+        hmac_key = tag->getString("hmac_key", "recaptchaDefaultKey");
+
+        // Setup SQL provider (recommended posgresql on very large networks)
+        std::string dbid = tag->getString("dbid", "default");
+        sql.SetProvider("SQL/" + dbid);
+
+        if (!sql)
+        {
+            throw ModuleException(this, INSP_FORMAT("*** reCAPTCHA: Could not find SQL provider with id '{}'.", dbid));
+        }
+    }
+
+    ModResult OnUserPreJoin(LocalUser* user, Channel* chan, const std::string& cname, std::string& privs, const std::string& keygiven, bool override) override
+    {
+        if (user->IsOper())
+            return MOD_RES_PASSTHRU;
+
+        // NickServ will help ya to not do the recaptcha
+        if (accountapi && accountapi->GetAccountName(user))
+        {     // we +r we good
+            user->WriteNotice("*** reCAPTCHA: NickServ account verified. You may join channels.");
+            return MOD_RES_PASSTHRU;
+        }
+
+        auto& tag = ServerInstance->Config->ConfValue("captchaconfig");
+        std::set<std::string> whitelist_channels;
+        std::stringstream chanstream(tag->getString("whitelistchans"));
+        std::string channel_name;
+        while (std::getline(chanstream, channel_name, ','))
+            whitelist_channels.insert(channel_name);
+
+        if (whitelist_channels.find(cname) != whitelist_channels.end())
+            return MOD_RES_PASSTHRU;
+
+        std::set<int> whitelist_ports;
+        std::stringstream portstream(tag->getString("whitelistports"));
+        std::string port_str;
+        while (std::getline(portstream, port_str, ','))
+            whitelist_ports.insert(std::stoi(port_str));
+
+        if (whitelist_ports.find(user->server_sa.port()) != whitelist_ports.end())
+            return MOD_RES_PASSTHRU;
+
+        if (captcha_verified.Get(user))
+            return MOD_RES_PASSTHRU;
+
+        NotifyUserToVerify(user);
+        return MOD_RES_DENY;
+    }
+
+    void ValidateToken(User* user, const std::string& token)
+    {
+        if (!sql)
+        {
+            user->WriteNotice("*** reCAPTCHA: SQL database is not available.");
+            return;
+        }
+
+        std::string formatted_query = INSP_FORMAT(
+            "SELECT COUNT(*) FROM recaptcha_app_verificationtoken WHERE token = '{}' AND created_at + INTERVAL '30 minutes' > NOW()",
+            token);
```

Comment:
This contains a SQL vulnerability. Someone could run e.g. `/VERIFY :foo'; DROP TABLE recaptcha_app_verificationtoken; ---` to drop your entire table.

Instead of formatting here you should use the ParamList or ParamMap overloads of Submit to format this safely.